go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -523,13 +523,62 @@ func (npw *nodePortWatcher) UpdateService(old, new *kapi.Service) {
 	}
 }
 
+// deleteConntrackForServiceVIP deletes the conntrack entries for the provided svcVIP:svcPort by comparing them to ConntrackOrigDstIP:ConntrackOrigDstPort
+func deleteConntrackForServiceVIP(svcVIPs []string, svcPorts []kapi.ServicePort, ns, name string) error {
+	for _, svcVIP := range svcVIPs {
+		for _, svcPort := range svcPorts {
+			err := deleteConntrack(svcVIP, svcPort.Port, svcPort.Protocol, true)
+			if err != nil {
+				return fmt.Errorf("failed to delete conntrack entry for service %s/%s with svcVIP %s, svcPort %d, protocol %s: %v",
+					ns, name, svcVIP, svcPort.Port, svcPort.Protocol, err)
+			}
+		}
+	}
+	return nil
+}
+
+// deleteConntrackForService deletes the conntrack entries corresponding to the service VIPs of the provided service
+func (npw *nodePortWatcher) deleteConntrackForService(service *kapi.Service) error {
+	// remove conntrack entries for LB VIPs and External IPs
+	externalIPs := util.GetExternalAndLBIPs(service)
+	if err := deleteConntrackForServiceVIP(externalIPs, service.Spec.Ports, service.Namespace, service.Name); err != nil {
+		return err
+	}
+	if util.ServiceTypeHasNodePort(service) {
+		// remove conntrack entries for NodePorts
+		nodeIPs := &npw.nodeIPManager.addresses
```

Comment:
this doesn't look thread safe. nodeIPManager uses a lock to control reading/writing these addresses. It looks like there is a pattern in the code directly reading this which we need to fix like:
		hasLocalHostNetworkEp := hasLocalHostNetworkEndpoints(ep, &npw.nodeIPManager.addresses)

We need a get method in the addresses I think to return a copy of the set.

Reply:
oh wow good catch, I must admit I didn't ever consider this problem...